### PR TITLE
Added support for dot notation with data elements tokens. Removed support for most magical tokens.

### PR DESCRIPTION
**Pull request by Aaronius**
Currently we have magically delicious tokens where you can use dot notation to access nested properties of an object. For example, when you have a rule where the event type is "click" and you're setting an evar value for analytics, in the textfield you can enter %event.target.id% and the evar will be set to the ID of the element that was clicked. I call these magically delicious tokens because there's no actual data element called `event`.

We support dot notation for the following:

```
%this.whatever%
%event.whatever%
%target.whatever%
%param.whatever%
```

The problem is that we don't support it for any data elements that users create. For example, if you create a data element called `foo`, you can't do this in the evar textfield: `%foo.whatever%`

We need to support dot notation for actual data elements that users create.

This affects migration. Currently in DTM we allow users to create data elements with periods in their name. These won't work when we add support for dot notation since any periods would be interpreted as trying to access nested properties.

------------------

Currently when filling out a text field that supports data elements (for example, populating an analytics evar), we allow you to use certain magical tokens even when you haven't created any data elements. Here are the magical tokens that we support:

```
%URI%
%uri%
%protocol%
%hostname%
%this.whatever%
%event.whatever%
%target.whatever%
%window.whatever%
%param.whatever%
%rand4% (where 4 means create a random number with 4 digits. The number can be anything)
```

I've talked with Corey Spencer and we want to only support the following magical tokens (for backward compatibility reasons):

```
%this.whatever%
%event.whatever%
%target.whatever%
```

For the other magical tokens, the user should instead create appropriate data elements. If any of these things are commonly used, we can create data element types for them in the Core extension to help out. For example, we could create a "random number" data element type where the user inputs how many digits they want and we generate a number for them.

### CODE REVIEW

**Comment by jbradley-sprint** [comment]
We use dots in our data element names as a sort of psuedo-namespacing.  Things like locator.state or locator.zip.  They do not refer to any actual locator object - just a way of grouping elements.  When migrating from DTM to Launch, will we have to modify those names and all rules that reference them, or will they still be supported?

I was not aware of the "magically delicious" tokens until I read this PR.  You learn something new every day!

**Comment by Aaronius** [comment]
@jbradley-sprint Thanks for asking. Yes, you would need to modify your data element names and the rules that reference them. Question for you: would you prefer _not_ adding support for accessing properties using dot notation if it meant you didn't have to modify your data element names and the rules that reference them?

**Comment by jbradley-sprint** [comment]
@Aaronius Selfishly speaking, having to modify all those elements and rules will be a barrier to - or at least a serious drag on the pace of - migration for some of our properties.  Many of our data elements were named (by various folks, including Adobe consulting) with dots.  There was nothing to stop it, and as far as I know, no documentation that discouraged it.

I see the benefit of maintain support for the three token you listed above, even though it is not something we ever used (or, as I mentioned, even knew about).  If you take away the ability to include dots in data element names, I would ask that some sort of tool be available to make it easier to change the names and the rules and javascript code that reference them.  For example, the property for the legacy Boost Mobile site has over 200 data elements with dots in their name (not my doing!)

**Comment by Aaronius** [comment]
Thank you. I'll let the product managers know and they can decide how they would like to move forward.

**Comment by Aaronius** [comment]
@jbradley-sprint I talked it over with Corey and we decided to remove the ability to access nested properties on data element values with using the `%` tokens. This means periods will continue to be supported in data element names and you won't have to do any renaming during migration. Thanks again for the feedback.

File: src/getVar.js [hunk 1]
```diff
@@ -43,8 +41,7 @@ var specialPropertyAccessors = {
  * @param supportSpecial
  * @returns {*}
  */
-var getObjectProperty = function(host, path, supportSpecial) {
-  var propChain = path.split('.');
+var getObjectProperty = function(host, propChain, supportSpecial) {
   var value = host;
   var attrMatch;
   for (var i = 0, len = propChain.length; i < len; i++) {
```
Comment (dorgan) [inline src/getVar.js]: Do you have to support IE older than version 9?  Using Array.forEach() make for better readability and less variable assignlengths/cleaner code.  changing this to a forEach you get rid of `var i = 0, len = propChain.length`

**Pull request merged**